### PR TITLE
refactor scheduler client using grpc  consistent hashring balancer

### PR DESCRIPTION
**Pull request by bigerous**
## Motivation and Context
Refactor scheduler client grpc consistent hashring using grpc balancer.
- The scheduler client's ClientConn will use d7y://scheduler as target address. A global resolver with DynConfig will resolve this address .
- Resovler will register an observer to be notified the update of DynConfig, and get all scheduler addresses.
- Balancer manage all SubConn to each scheduler. When resolver tell balancer different scheduler addresses, balancer will generate new hashring picker.
- SubConn state is Ready only when it is picked according to consistent hashring
- If a SubConn connect failed exceed 3 times, it will be treated as bad node, and be removed from hashring, it will still try to connect to scheduler with backoff strategy until success or it is removed by resolver.
- When grpc call, picker and balancer will cooperate to offer a Ready SubConn.
- SubConn state will be transformed to Idle 5 seconds after grpc call is done, to reduce pressure of scheduler.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: pkg/consistent/connection.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+func (cp *conPool) Watch() {
+	ticker := time.NewTicker(time.Second)
+	for {
+		select {
+		case <-ticker.C:
+			for _, scw := range cp.Values() {
+				if scw.NeedGC(cp.gcInterval) {
+					scw.UpdateState(TemporarilyClosed)
+					scw.Disconnect()
```
Comment (jim3ma) [inline pkg/consistent/connection.go:173]: Should also remove sc in conPool.cons and conPool.subCons ?

File: pkg/consistent/connection.go [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+func (scw *subConnWrapper) Disconnect() {
+	scw.cc.RemoveSubConn(scw.sc)
+}
+
+func (scw *subConnWrapper) Close() {
+	scw.cc.RemoveSubConn(scw.sc)
+}
+
+type conPool struct {
```
Comment (jim3ma) [inline pkg/consistent/connection.go:141]: `conPool ` -> `connPool `

File: pkg/consistent/picker.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		nodeKey, ok = p.hashRing.GetNode(string(key))
+		if !ok {
+			return balancer.PickResult{}, ErrBalancerHashring
+		}
+		subConn, ok = p.subs.Get(nodeKey)
+		if !ok {
+			return balancer.PickResult{}, ErrBalancerHashring
+		}
+		p.Debugf("grpc call of %s,hash_key %v, pick %s,", info.FullMethodName, key, nodeKey)
+		logger.WithTaskID(key).Debugf("grpc call of %s,hash_key %v, pick %s,", info.FullMethodName, key, nodeKey)
```
Comment (jim3ma) [inline pkg/consistent/picker.go:74]: `,hash_key` -> `, hash_key`
add space after `,`

File: pkg/consistent/picker.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	} else {
+		subConn, ok = p.subs.GetRandom()
+		if !ok {
+			return balancer.PickResult{}, ErrBalancerHashring
+		}
+		p.Debugf("grpc call of %s,no hash key, random pick %s,", info.FullMethodName, nodeKey)
```
Comment (jim3ma) [inline pkg/consistent/picker.go:80]: Print warning log ?

File: pkg/rpc/scheduler/client/client.go [hunk 5]
```diff
@@ -20,20 +20,32 @@ package client
...
+	grpc_prometheus "github.com/grpc-ecosystem/go-grpc-prometheus"
 	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/credentials/insecure"
 
 	commonv1 "d7y.io/api/pkg/apis/common/v1"
 	schedulerv1 "d7y.io/api/pkg/apis/scheduler/v1"
 
 	logger "d7y.io/dragonfly/v2/internal/dflog"
+	consistent2 "d7y.io/dragonfly/v2/pkg/consistent"
```
Comment (jim3ma) [inline pkg/rpc/scheduler/client/client.go:36]: Why use package alias `consistent2 ` ?

**Review by jim3ma — CHANGES_REQUESTED** [review]

File: pkg/consistent/dyn_config.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
+/*
```
Comment (gaius-qi) [inline pkg/consistent/dyn_config.go:1]: Rename `dyn_config.go` to `dynconfig.go`, refer to `internal/dynconfig`.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: pkg/consistent/connection.go (hunk 1 above)
Comment (bigerous) [reply pkg/consistent/connection.go:173]: remove operation is done in `balancer.UpdateSubConnState` when subConn state become ShutDown

**Pull request closed without merging**